src/System.Private.ServiceModel/src/Internals/System/Runtime/TaskHelpers.cs

```diff
@@ -180,26 +180,64 @@ public static void ToApmEnd(this IAsyncResult iar)
         }
 
         // Task.GetAwaiter().GetResult() calls an internal variant of Wait() which doesn't wrap exceptions in
-        // an AggregateException.
+        // an AggregateException. It does spinwait so if it's expected that the Task isn't about to complete,
+        // then use the NoSpin variant.
         public static void WaitForCompletion(this Task task)
         {
             task.GetAwaiter().GetResult();
         }
 
+        // If the task is about to complete, this method will be more expensive than the regular method as it
+        // always causes a WaitHandle to be allocated. If it is expected that the task will take longer than
+        // the time of a spin wait, then a WaitHandle will be allocated anyway and this method avoids the CPU
+        // cost of the spin wait.
+        public static void WaitForCompletionNoSpin(this Task task)
+        {
+            if (!task.IsCompleted)
+            {
+                ((IAsyncResult) task).AsyncWaitHandle.WaitOne();
```

Comment:
Just checking -- this is logically an infinite wait  (which of course, so it GetResult()).  Will this wait handle be set by if the task is cancelled?